### PR TITLE
drivers: flash: stm32 ospi: extend memory map support for other modes

### PR DESCRIPTION
**Pull request by GeorgeCGV**
Extends memory map support with quad, dual, and spi modes.

* Allow custom write opcode and adapt used lines based on read opcode.
  That allows memory map mode to be used with Winbond and other vendors.
* Add more clarity to the error log output to differentiate the error's root cause.
* Replace `LOG_INF` with `LOG_DBG` to signal about memory map mode being enabled.

### CODE REVIEW

**Review by FRASTM — COMMENTED** [review]
do you confirm that set memory_mapped is still funtionnal with a b_u585i_iot02a or stm32h7b3i

File: drivers/flash/flash_stm32_ospi.c [hunk 1]
```diff
@@ -987,75 +987,71 @@ static int stm32_ospi_set_memorymap(const struct device *dev)
...
+		case JESD216_MODE_144:
+			s_command.InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
+			s_command.AddressMode = HAL_OSPI_ADDRESS_4_LINES;
+			s_command.DataMode = HAL_OSPI_DATA_4_LINES;
+			break;
+		default:
+			/* Use lines based on data_mode set in ospi_prepare_cmd */
+			break;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (GeorgeCGV) [reply drivers/flash/flash_stm32_ospi.c:1033]: ?

The `ospi_prepare_cmd` contains: https://github.com/zephyrproject-rtos/zephyr/blob/740f69667a3f9ab6aad873721a4beabeea30572d/drivers/flash/flash_stm32_ospi.c#L324-L349

For the `OSPI_OPI_MODE` (`dev_cfg->data_mode`) that yields

```
s_command.InstructionMode = HAL_OSPI_INSTRUCTION_8_LINES
s_command.AddressMode = HAL_OSPI_ADDRESS_8_LINES
s_command.DataMode = HAL_OSPI_DATA_8_LINES
```

while the `write_opcode` would be `SPI_NOR_OCMD_PAGE_PRG` set in `spi_nor_process_bfp` for `OSPI_OPI_MODE` if DT doesn't set `writeoc`.

**Comment by GeorgeCGV** [comment]
> do you confirm that set memory_mapped is still funtionnal with a b_u585i_iot02a or stm32h7b3i

that would be hard to do as I don't have mentioned board. So far the memap mode is used on a custom board in quad mode against macronix and winbond nor-flashes. However, no writing is involved. But memap write configuration in this PR is pretty much the same as standard ospi write function.

**Comment by MaureenHelm** [comment]
@erwango @FRASTM please take a look

**Review by FRASTM — APPROVED** [review]

**Comment by FRASTM** [comment]
Tested on b_u585i_iot02a

**Review by erwango — APPROVED** [review]

**Pull request merged**